### DISCUSSION TITLE
How to create a function accepting a closure ?

### DISCUSSION THREAD

**Discussion by buixor** [antonmedv/expr#392] [General]
Hello,

I would like to know how to add a function accepting closure as an argument.

For example, I want to create a function `flatten( something, { #.some_field })`:
 - `something` would be a `[]map[string]string`
 - and it would return a `[]string` containing the `some_field` values found in the `something` array

To achieve this, I believe closure would be the best way. Is it something that can be done ? (I didn't find other examples except the builtins all/one/etc. but I don't see how to transpose this to a function I could expose).

Thanks!

**Comment by antonmedv** [comment]
You can just use this:
```
map(something, .some_field)
```
Brackets `{` and `}`, as well as `#` char is optional.

**Reply by buixor** [reply]
Hello,

Sorry, the flatten was just an example.

My question is really how/if I can create a function that accepts closure argument :smile: 

Thanks and sorry if the question was unclear

**Reply by antonmedv** [reply]
This is not possible.  Those predicates (all, none, map, filter, etc) and closures are tight together. Think of them as language keywords (like `for`, `while`, `do {...} while`). 

It may be cool to add some API to allow to use closures somehow.

What is your use case? What kind of problem you trying to solve?

**Reply by Arttii** [reply]
I was looking at the same today. My specific usecause is for a custom filter/count/sum where I want the closure to be a predicate pushdown to the left-hand side, which is a function, that does some sql queries. I was wondering what the most ergonomic way to implement this would be. 

I primarily want to avoid moving data unnecessarily.
